distributed/protocol/highlevelgraph.py

```diff
@@ -130,12 +130,17 @@ def _materialized_layer_unpack(state, dsk, dependencies, annotations):
         dependencies[k] = list(set(dependencies.get(k, ())) | set(v))
 
     if state["annotations"]:
-        annotations.update(
-            Layer.expand_annotations(state["annotations"], state["dsk"].keys())
-        )
+        expanded = Layer.expand_annotations(state["annotations"], state["dsk"].keys())
+        new_annotations = {}
+        for k, v in expanded.items():
+            if isinstance(v, dict) and k in annotations:
+                new_annotations[k] = merge(annotations[k], v)
+            else:
+                new_annotations[k] = v
+        annotations.update(new_annotations)
```

Comment:
I understand this change provides backwards compatibility for the existing per-key Client.{submit, persist} kwarg taxonomy (retries, priority, workers)?

The functionality in **_materialized_layer_unpack** mirrors [Layer.\_\_dask_distributed_unpack\_\_](https://github.com/dask/dask/blob/d05508163086a0d09c65cb7d91a3aaab35e79a2e/dask/highlevelgraph.py#L224-L231) and it all it's subclasses. They also need to expand and unpack annotations.

I wonder if it would be better to place this functionality in Layer.expand_annotations so that it automatically is applied through the Layer class hierarchy? It would require a sister dask PR though.

Reply:
Thanks for pointing this out. This is actually not for backwards compatibility of per-key/collection annotations, but instead is needed to handle the case where different HLG Layers have different retries/priority/workers.

For example, if we are unpacking a two-layer HLG, and Layer A has `retries:2` while Layer B has `retries:4`, we need to make sure that the unpacking keeps both of those annotations. The current implementation has a shallow merge, which makes later-unpacked layers overwrite the annotations of previously-unpacked layers. To make the above case work, we need a bit deeper of a merge.

For the same reason, I also think this logic needs to be one step above `Layer.expand_annotations`.

Reply:
Yes, you're right. Thanks for correcting my understanding.

```python
new_annotations[k] = {**annotations[k], **v}
```

may be faster than tlz.merge now that dask supports >= Python 3.6

Reply:
cc'ing @madsbk

Reply:
Thanks @madsbk, this is a great point. There is also the `Layer.expand_annotations` function called in the default implementation, which it may make sense to remove for the same reasons. @sjperkins are you aware of any major downsides to doing this?

Reply:
@ian-r-rose @madsbk I do have some objections around [low code cohesion](https://en.wikipedia.org/wiki/Cohesion_(computer_science)), but I'm trying to provide some designs to rectify this. I'll try answer in more detail today.

Reply:
I'm concerned about [low code cohesion](https://en.wikipedia.org/wiki/Cohesion_(computer_science)) in the sense that functionality for serialising and deserialising layers is spread across dask and dask distributed  (_materialized_layer_pack and _materialized_layer_unpack) in particular. Moving annotation expansion into dask distributed further exacerbates this trend.

`__dask_distributed_pack__` and `__dask_distributed_unpack__` appear to describe an API for serialising/deserialising layers similar to pickle, but `__dask_distributed_unpack__` does not reconstruct the original Layer, instead returning a low-level dask graph and associated dependencies for ingest by the scheduler.

I suggest:

1. Modifying `__dask_distributed_pack__` to return a `(cls, args, kwargs)` tuple (similar to the `__reduce__` protocol) for encoding by msgpack instead of a state dict . `__dask_distributed_pack__` could be renamed to `__dask_distributed_reduce__`.

2. This can be used to reconstruct the Layer server-side:

    ```python
   # client-side
   state = mspack.pack(layer.__dask_distributed_reduce__(client))
   # server-side
    cls, args, kwargs = msgpack.unpack(state)
    layer = cls(*args, **kwargs)
    ```

3. Replace classmethod  `__dask_distributed_unpack__(cls, state)` with instance method `__dask_distributed_unpack__(self)` which provides the raw `dsk` and `deps` dicts.

4.  Replace `_materalized_layer_pack` and `_materialized_layer_unpack` in dask distributed with a `class MaterializedLayer(Layer)` in dask that encapsulates the same functionaliy.

With this design, annotations can be packed in `(cls, args, kwargs)` automatically by Layer subclass implementers calling super() in `__dask_distributed_reduce__` (see extended example below).

Also MaterializedLayer gets placed in dask so there's no split across dask distributed.

This design prevents further low cohesion across the code base.

The code below illustrates most of these ideas:


```python
stringify = lambda x: x

class Layer:
    def __init__(self, annotations=None):
        self.annotations = annotations

    def __dask_distributed_reduce__(self, client):
        """
        Returns
        -------
        cls : class
        *args : Arguments for constructing a layer
        *kwargs : Keyword arguments for constructing a layer
        """
        return Layer, (), {"annotations": self.pack_annotations()}

    def __dask_distributed_unpack__(self):
        """
        Returns
        -------
        dsk : dict
        deps : dict
        """
        raise {stringify(k): v for k, v in self.items()}, {}


    def pack_annotations(self):
        """ This should actually do annotation packing """
        return self.annotations


    def materialize(self, all_keys, known_key_dependencies, client, client_keys):
        """ Implement _materialized_layer_pack """
        return MaterializedLayer({"dask": "graph"}, {"deps": "keys"},
                                 annotations=self.annotations)


class BasicLayer(Layer):
    def __init__(self, mapping, annotations=None):
        super().__init__(annotations=annotations)
        self.mapping = mapping

    def __dask_distributed_reduce__(self, client):
        _, args, kwargs = super().__dask_distributed_reduce__(client)
        assert not args
        return self.__class__, (self.mapping,), kwargs

    def __dask_distributed_unpack__(self):
        dsk, deps = super().__dask_distributed_unpack__()
        # update deps
        return dsk, deps


class Blockwise(Layer):
    def __init__(self, output, output_indices, dsk, indices,
                 numblocks, concatenate, new_axes, output_blocks,
                 annotations=None):
        super().__init__(annnotations=annotations)
        self.output = output
        self.output_indices = output_indices
        self.dsk = dsk
        self.indices = indices
        self.numblocks
        self.concatenate = concatenate
        self.new_axes = new_axes
        self.output_blocks = output_blocks

    def __dask_distributed_reduce__(self, client):
        _, args, kwargs = super().__dask_distributed_reduce__(client)
        assert not args
        args = (self.output, self.output_indices, self.dsk, self.indices,
                self.numblocks, self.concatenate, self.new_axes,
                self.output_blocks)
        return Blockwise, args, kwargs

    def __dask_distributed_unpack__(self):
        raw, raw_deps = make_blockwise_graph()
        dsk = {stringify(k): v for k, v in raw.items()}
        deps = {stringify(k): v for k, v in raw_deps.items()}
        return dsk, deps


class MaterializedLayer(Layer):
    def __init__(self, dsk, deps, annotations=None):
        super().__init__(annotations=annotations)
        self.dsk = dsk
        self.deps = deps

    def __dask_distributed_reduce__(self, client):
        cls, args, kwargs = super().__dask_distributed_reduce__(client)

        # Update args, kwargs etc.
        assert not args
        args = args + (self, dsk, self.deps)
        return MaterializedLayer, args, kwargs

    def __dask_distributed_unpack__(self):
        dsk = {stringify(k): v for k, v in self.dsk.items()}
        deps = {stringify(k): v for k, v in self.deps.items()}
        return dsk, deps


def encode_layer(client, layers):
    """ Client-side """
    return [msgpack.pack(l.__dask_distributed_reduce__(client)) for l in layers]


def decode_layers(layers):
    """ Server side """"
    decoded_layers = []

    for layer in layers:
        cls, args, kwargs = msgpack.unpack(layer)
        decoded_layers.append(cls(*args, **kwargs))

    return decode_layers

def reify_layers(layers):
    """ Server side """
    dsk = {}
    deps = {}
    annotations = {}

    for layer in layers:
        layer_dsk, layer_deps = layer.__dask_distributed_unpack__()
        dsk.update(layer_dsk)
        deps.update(layer_deps)
        annotations.update(Layer.expand_annotations(layer.annotations, layer.keys()))

    return dsk, deps, annotations
```

Reply:
Initially, we decided to implement the functionality that depended on Distributed in Distributed in order to avoid importing distributed in Dask. However, since we have to do that anyway e.g. in Blockwise I agree that it makes sense to move everything to Dask. I also agree that implementing the default pack/unpack in `Layer` makes a lot of sense. 

However, I don't think we need to change the basic logical flow and because of the asymmetric nature of `__dask_distributed_pack__` and `__dask_distributed_unpack__` and because we have very [specific requirement for `__dask_distributed_pack__`](https://github.com/dask/dask/blob/30c9a3a34240e08cccd7e38b115141661ca198cf/dask/highlevelgraph.py#L202-L207), I don't we should rename it to `__dask_distributed_reduce__`. 

Also notice that:
`__dask_distributed_unpack__` needs `dsk` and `dependencies` from the preceding layers.
`__dask_distributed_reduce__` for security reasons we cannot de-serializes class/modules directly.

Reply:
> However, I don't think we need to change the basic logical flow and because of the asymmetric nature of `__dask_distributed_pack__` and `__dask_distributed_unpack__` and because we have very specific requirement for `__dask_distributed_pack__`, I don't we should rename it to `__dask_distributed_reduce__`.

I read the requirements and understand this better now.

> Also notice that:
`__dask_distributed_unpack__` needs dsk and dependencies from the preceding layers.

Ah, I see. Then, to separate annotation expansion from `__dask_distributed_unpack__` I would say that `__dask_distributed_unpack__` should not modfiy the dsk and deps arguments, but instead construct layer specific dsk and deps for return (Layer.expand_annotations needs to know the keys for the associated layer).

I almost might suggest modifying `__dask_distributed_unpack__` to return a dict for future-proofing return of any additional state required on the scheduler. I think the dict looks a bit aesthetically ugly, but I can't think of anything better off-hand. Thoughts?

```python
class SomeLayer(Layer):
    def __dask_distributed_unpack__(cls, state, dsk, deps):
        layer_dsk = {... using state }
        layer_deps = {... using state and dsk and deps }

        return {"dsk": layer_dsk, "deps": layer_deps, "keys": layer_dsk.keys()}

def unpack(layer_state):
    dsk  = {}
    deps = {}
    annotations = {}

    for state in layer_state:
        layer_cls = state["class"]
        state = state["state"]
        data = layer_cls.__dask_distributed_unpack(cls, state, dsk, deps)
        dsk.update(data["dsk"])
        deps.update(data["deps"])
        # Layer.expand_annotations must expand over the layer keys only
        annotations.update(Layer.expand_annotations(state["annotations"], data["dsk"].keys())
    
   return dsk, deps, annotations
```

`__dask_distributed_reduce__` for security reasons we cannot de-serializes class/modules directly.

:+1:

Reply:
An alternative where `__dask_distributed_unpack__` returns a dict and its possible to defer to the superclass for annotation unpacking

```python
class Layer:
    def __dask_distributed_pack__(self, client):
        return {"annotations": self.pack_annotations}

    def __dask_distributed_unpack__(cls, state, dsk, deps):
        try:
            annotations = state["annotations"]
        except KeyError:
            return {}

        try:
            layer_dsk = state["dsk"]
        except KeyError:
            raise ValueError("dsk must be in state to expand annotations over layer keys")
        else:
            keys = layer_dsk.keys()

        return {"annotations": cls.expand_annotations(annotations, keys)}

class SomeLayer(Layer):
    def __dask_distributed_unpack__(cls, state, dsk, deps):
        layer_dsk = {... using state }
        layer_deps = {... using state and dsk and deps }

        # Defer to superclass for automatic layer unpacking.
        return {"dsk": layer_dsk,
                    "deps": layer_deps":
                    "keys": layer_dsk.keys(),
                   **super().__dask_distributed_unpack__({"dsk": layer_dsk, **state}, dsk, deps)}


def unpack(layer_state):
    dsk  = {}
    deps = {}
    annotations = {}

    for state in layer_state:
        layer_cls = state["class"]
        state = state["state"]
        data = layer_cls.__dask_distributed_unpack(cls, state, dsk, deps)
        dsk.update(data["dsk"])
        deps.update(data["deps"])
        annotations.update(data["annotations"])
    
   return dsk, deps, annotations
```

Reply:
Thanks for the detailed discussion @sjperkins and @madsbk. I've tried to do something like a compromise solution in dask/dask#7102, and I'd love to hear your thoughts on it. It takes a somewhat minimal approach to API changes (so `__dask_distributed_unpack__()` still takes the annotations), and avoids further imports of `distributed` there. Overall the strategy I'm trying out is

1. Move the default implementation of layer unpacking into `Layer`.
2. Move the logic of doing a deeper merge on annotations into a new `Layer.merge_annotations`, similar to `Layer.expand_annotations`. I am not overly attached to this division of responsibility, so if there are suggestions for how to make it cleaner, I'd be happy to hear them.
3. Ensure that `Layer`, `Blockwise`, and `ShuffleLayer` all unpack annotations properly (previously the latter two failed).

Reply:
LGTM @ian-r-rose but I think @sjperkins has a point, we should unify everything in Dask and change the API such that layers can delegate the annotation unpacking to `Layer.__dask_distributed_unpack__()`.

I suggest that we split this work into multiple PRs. Let's rollback this PR to the state that didn't require any significant change to the API and get it merged. A followup PR can then modify the API based on the design thoughts we have been discussing here. I will be happy to give that a try.

Reply:
Thanks for the discussion @madsbk and @ian-r-rose . I agree with moving Layer API changes forward to a new PR, so don't let my concerns hold this one up. LGTM

Reply:
@madsbk I'd be happy to have this be across multiple PRs. Can you specify what API changes you would want to roll back? This now depends on dask/dask#7102, but that is somewhat more of a bugfix than a major API change.

I could bring the default implementation of `Layer.__dask_distributed_unpack__()` back here and then xfail the test that catches dask/dask#7102 , if that's what you mean.

Reply:
Sorry my bad, I thought you changed the API in the last commit. I am totally fine with this PR+dask/dask#7102